src/primitives/hrp.rs

```diff
@@ -0,0 +1,406 @@
+    /// uppercase ASCII `char`s. For lowercase bytes see [`Self::lowercase_byte_iter`]
+    pub fn byte_iter(&self) -> ByteIter { ByteIter { iter: self.buf.iter().take(self.size) } }
+
+    /// Creates a character iterator over the ASCII characters of this HRP.
+    ///
+    /// If an uppercase HRP was parsed during object construction then this iterator will yield
+    /// uppercase ASCII `char`s. For lowercase bytes see [`Self::lowercase_char_iter`].
+    pub fn char_iter(&self) -> CharIter { CharIter { iter: self.byte_iter() } }
+
+    /// Creates a lowercase iterator over the byte values (ASCII characters) of this HRP.
+    pub fn lowercase_byte_iter(&self) -> LowercaseByteIter {
+        LowercaseByteIter { iter: self.byte_iter() }
+    }
+
+    /// Creates a lowercase character iterator over the ASCII characters of this HRP.
+    pub fn lowercase_char_iter(&self) -> LowercaseCharIter {
+        LowercaseCharIter { iter: self.lowercase_byte_iter() }
+    }
+
+    /// Returns the length (number of characters) of the human-readable part.
+    ///
+    /// Guaranteed to be between 1 and 83 inclusive.
+    pub fn len(&self) -> usize { self.size }
+
+    /// Always false, the human-readable part is guaranteed to be between 1-83 characters.
+    pub fn is_empty(&self) -> bool { false }
+}
+
+/// Displays the human-readable part.
+///
+/// If an uppercase HRP was parsed during object construction then the returned string will be
+/// in uppercase also. For a lowercase string see [`Self::to_lowercase`].
+impl fmt::Display for Hrp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for c in self.char_iter() {
+            f.write_char(c)?;
+        }
+        Ok(())
+    }
+}
+
+/// Case insensitive comparison.
+impl Ord for Hrp {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.lowercase_byte_iter().cmp(other.lowercase_byte_iter())
+    }
+}
+
+/// Case insensitive comparison.
+impl PartialOrd for Hrp {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> { Some(self.cmp(other)) }
+}
+
+/// Case insensitive comparison.
+impl PartialEq for Hrp {
+    fn eq(&self, other: &Self) -> bool {
+        self.lowercase_byte_iter().eq(other.lowercase_byte_iter())
+    }
+}
+
+impl Eq for Hrp {}
+
+impl core::hash::Hash for Hrp {
+    fn hash<H: core::hash::Hasher>(&self, h: &mut H) { self.buf.hash(h) }
+}
+
+/// Iterator over bytes (ASCII values) of the human-readable part.
+///
+/// ASCII byte values as they were initially parsed (i.e., in the original case).
+pub struct ByteIter<'b> {
+    iter: iter::Take<slice::Iter<'b, u8>>,
+}
+
+impl<'b> Iterator for ByteIter<'b> {
+    type Item = u8;
+    fn next(&mut self) -> Option<u8> { self.iter.next().cloned() }
+
+    fn size_hint(&self) -> (usize, Option<usize>) { (self.len(), Some(self.len())) }
+}
+
+impl<'b> ExactSizeIterator for ByteIter<'b> {
+    fn len(&self) -> usize { self.iter.len() }
+}
+
+impl<'b> FusedIterator for ByteIter<'b> {}
```

Comment:
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

I guess we ought to implement `DoubleEndedIterator` as well for good measure.

Reply:
Done for each iterator and tested (noticed I missed on FusedIterator impl also).